### PR TITLE
BFD-1925: add CI workflows for fda drug code

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-1925](https://jira.cms.gov/browse/BFD-1925)

**User Story or Bug Summary:**

User Story
As a developer I want the FDA library to build successfully in github actions, in jenkins, so it can be uploaded to AWS Code Artifact
 

AC

1. Add the Jenkinsfile to the FDA project
2. Make sure the project builds in GitHub Actions and in Jenkins. (SIA is probably needed for this)
3. The FDA LIbrary is uploaded to AWS Code Artifact 
4. The jenkinsfile is used by jenkins
5. AWS Code Artifact needs to be in terraform (SIA is probably needed for this). - this will be a separe PR by michael burling

---

### What Does This PR Do?

This adds a jenkins file and github actions for the new awa code artifactory

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* BFD Team had a meeting with Jason and Karl over the implications and any security concerns of this PR.  Furthermore we have entered a SIA into the process.  After talking with the security team the conclusion was we didn't have to do that for this change since we are using already AWS approved services.  The BFD team has updated the diagram in the SDD under section 2.1.4.3 for this change to our deployment pipeline and how the new jar file is being used.  

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-data-fda/Jenkinsfile
Comment (mackec1) [inline apps/bfd-data-fda/Jenkinsfile:29]: might be nice to have a comment that explain what the cron schedule results in

**Review by mackec1 — APPROVED** [review]
looks reasonable; had a comment that it might be nice to show what the cron sched manifests as

File: apps/bfd-data-fda/Jenkinsfile
Comment (cbrunefearless) [reply apps/bfd-data-fda/Jenkinsfile:29]: ill add

**Review by mackec1 — DISMISSED** [review]
Looks good; might need to leverage portions for stuff I'm doing!

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:116]: This was a poor diagnostic we were using which can be removed safely now. Also harmless to keep in. I'll leave it up to you.
```suggestion
```

File: .github/workflows/ci-java.yml
Comment (mjburling) [inline .github/workflows/ci-java.yml:128]:
```suggestion
          mvn --quiet --batch-mode --threads 1C verify
```

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:121]: A note here about using this style of templated string instead of re-using `$CODEARTIFACT_ENDPOINT` might be helpful. Baking the domain/repository into this string is a little more brittle. I'm still a little puzzled why the equivalent seemed to yield:

> Invalid syntax for alternative repository. Use "id::layout::url".

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:106]: A single-line comment explaining this conditional might be helpful, e.g. "Only build and deploy the artifact when changes are detected based on the sha256 hash" or similar.

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:2]: A brief, high-level overview of what this Jenkinsfile is for would be helpful, e.g.
```suggestion
/** 
 * Enables Conditional daily build of a latest bfd-data-fda SNAPSHOT artifact from master using the bfd-cbc-build container image.
 * Delivers to bfd-mgmt AWS CodeArtifact Maven Repository as appropriate.
 */
```

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:13]: Addition of that TODO is still warranted here. Can you re-add the TODO on resolving the container image rather than hard-coding it?

**Review by mjburling — CHANGES_REQUESTED** [review]
Thanks, @cbrunefearless! This looks pretty good, but there are just a few minor things to address before we merge–basically in-line documentation. What do you think?

File: apps/bfd-data-fda/Jenkinsfile
Comment (cbrunefearless) [reply apps/bfd-data-fda/Jenkinsfile]: done

File: apps/bfd-data-fda/Jenkinsfile
Comment (cbrunefearless) [reply apps/bfd-data-fda/Jenkinsfile]: done

File: .github/workflows/ci-java.yml
Comment (cbrunefearless) [reply .github/workflows/ci-java.yml:128]: gone

File: apps/bfd-data-fda/Jenkinsfile
Comment (cbrunefearless) [reply apps/bfd-data-fda/Jenkinsfile:128]: done

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [reply apps/bfd-data-fda/Jenkinsfile:133]: Revisiting this... it might be fine as it is, but leaving _something_ behind for the next engineer to understand why we did _this_ instead of _that_ might be helpful, limited to a single line admitting that we just didn't go much further in trying to understand why the Jenkins didn't accept the cli-provided `id::layout::url` format with one interpolated value over another would be sufficient, but I'll leave that up to you.

**Review by mackec1 — APPROVED** [review]

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:117]:
```suggestion
                        // Check whether the parameter is empty or the sha value is not equal to the new sha value, if either 
```

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:50]: I think this line can be omitted now.

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:7]:
```suggestion
 * this is the case, then it delivers to bfd-mgmt AWS CodeArtifact Maven Repository as appropriate.
```

File: apps/bfd-data-fda/Jenkinsfile
Comment (mjburling) [inline apps/bfd-data-fda/Jenkinsfile:99]:
```suggestion
                      // Download and Calculate FDA DrugCode Sha256 from the external source
```

**Review by mjburling — CHANGES_REQUESTED** [review]
Just some minor typos, etc. >99.9% at this point.

**Review by mackec1 — DISMISSED** [review]
LGTM

File: .github/workflows/ci-java.yml
Comment (mjburling) [reply .github/workflows/ci-java.yml:128]: But it came back!

Comment (cbrunefearless) [reply .github/workflows/ci-java.yml]: took out

**Review by mackec1 — APPROVED** [review]

**Review by mjburling — APPROVED** [review]

**Pull request merged**
